### ISSUE TITLE
Output formatters: serializers (and transformers maybe)

### ISSUE THREAD

**Issue by jfexyz** [ferryjs/ferry#5]
One other layer we should think about is output formats. There are quite a few primary ones: [JSON API](http://jsonapi.org), [JSON HAL](https://github.com/mikekelly/hal_specification),  [JSON-LD](http://json-ld.org/), [etc](http://sookocheff.com/posts/2014-03-11-on-choosing-a-hypermedia-format/).

There's also possibly transformers, though this is less likely necessary in our case, and might be mitigated by hooks in resource handlers (though I could see using the default handlers, and just tweaking the format slightly). One source of inspiration here is [Fractal](http://fractal.thephpleague.com).

**Comment by desmondmorris** [comment]
@joshuajabbour  This is really interesting, I am not too familiar with either concept.  Would there be a world where both exist?  Can you give an example?

**Comment by jfexyz** [comment]
Yes, both definitely could exist. In Fractal, serializers are the general, non-resource-specific formats (e.g. JSON-HAL). Transformers are resource-specific, and mainly are for type coercion since PHP sucks at that, but can also be used to add properties to the result object, etc.

In our case, serializers will be an eventual must-have. Transformers may not be required, and may just be middleware that manipulates the result.
